### PR TITLE
Speedup 05: Retrieve unique detections in family and in `matched_filter`

### PR DESCRIPTION
**Pull request by flixha**
### What does this PR do?
- `core.match_filter.family._uniq`: **1.9x speedup**
  - retrieving the unique list of detections is quicker for many detections with `list(set)` (1.9x speedup for 43000 detections, fastest: 3.1 s), but 1.2x slower for small sets (e.g., 430 detections; 50 ms --> 27 ms).
- `core.match_filter.detect` - **1000x speed up for many calls to `family._uniq`**
  -  using `family._uniq` in a loop over all families is still rather slow with `_uniq`. Checking tuples of `(detection.id, detection.detect_time, detection.detect_val)` with `numpy.unique` and avoiding a loop is 1000x faster.  From 752 s to <1 s for 82000 detections.

### Why was it initiated?  Any relevant Issues?
Retrieving unique detections for `matched_filter`-run was getting slower than needed when there are a lot of detections.

This PR contributes to the summary issue in https://github.com/eqcorrscan/EQcorrscan/issues/522

### PR Checklist
- [X] `develop` base branch selected?
- [X] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests still pass.
~- [ ] Any new features or fixed regressions are be covered via new tests.~
~- [] Any new or changed features have are fully documented.~
- [X] Significant changes have been added to `CHANGES.md`.
~- [ ] First time contributors have added your name to `CONTRIBUTORS.md`.~

### CODE REVIEW

File: eqcorrscan/core/match_filter/family.py [hunk 1]
```diff
@@ -326,8 +326,7 @@ def _uniq(self):
         2
         """
         _detections = []
-        [_detections.append(d) for d in self.detections
-         if not _detections.count(d)]
+        _detections = list(set(self.detections))
```
Comment (calum-chamberlain) [inline eqcorrscan/core/match_filter/family.py:329]: Making a `set` of detections relies on them being hashable, which they are not (or should not be - they can and do change - I did accidentally make them hashable at one point). Therefore I don't think this method should work. If it isn't returning the following error:
```python
TypeError: unhashable type: 'Detection'
```
then it should!

File: eqcorrscan/core/match_filter/tribe.py [hunk 2]
```diff
@@ -606,7 +606,21 @@ def detect(self, stream, threshold, threshold_type, trig_int, plot=False,
...
+                        for det in family]
+                    # Retrieve the indices for the first occurrence of each
+                    # detection in the family (so only unique detections will
+                    # remain).
+                    uniq_det_tuples, uniq_det_indices = np.unique(
+                        det_tuples, return_index=True, axis=0)
+                    uniq_detections = []
+                    for uniq_det_index in uniq_det_indices:
+                        uniq_detections.append(family[uniq_det_index])
+                    family.detections = uniq_detections
```
Comment (calum-chamberlain) [inline eqcorrscan/core/match_filter/tribe.py:623]: This should be fine (for what we want, which is no multiple detections at the same time for the same template and hence family, this is ideal) - and a lot faster! Nice!

**Review by calum-chamberlain — CHANGES_REQUESTED** [review]
It looks like the changes to `_uniq` shouldn't work, and if they do then we should make sure they don't... Otherwise this is a really useful speedup!

File: eqcorrscan/core/match_filter/family.py (hunk 1 above)
Comment (flixha) [reply eqcorrscan/core/match_filter/family.py:329]: Thanks for pointing this out.. indeed I didn't get the error in any of my tests which is surprising considering that these errors came up in all tests on github. I'll revert this change here

**Review by calum-chamberlain — APPROVED** [review]
Golden, thanks!

**Pull request merged**